### PR TITLE
Fix context stack

### PR DESCRIPTION
**Pull request by rpgoldman**
Previously, it was possible to get the wrong type of `Context` subclass when using `Context.get_context()`.  The patch in the second commit fixes this error by searching the stack for values of the right type.  The first commit gives a test that illustrates the incorrect behavior.

If you wish to verify the error, fetch the branch for this PR, and then checkout the first commit only, and run `pymc3/tests/test_modelcontext.py`

I stumbled on this when vectorizing `sample_posterior_predictive`, because I used `modelcontext()` inside `with _DrawValuesContext():`, which does not happen in the current code base.

Nevertheless, I think that this is worth fixing now, because it's a bug waiting to catch us, and if the bug *does* manifest, then you can get back a `_DrawValuesContext` from a call to `modelcontext(None)`, and this bug is *not* guaranteed to be detected when you try to dereference a name inside what you *think* is a model, but is in fact a `_DrawValuesContext`.  Instead of an error, you just get back `None` when you should have gotten a variable, which is the worst kind of failure.

### CODE REVIEW

**Comment by ColCarroll** [comment]
this looks good to me -- seems like 3.5 is having typing problems

**Comment by rpgoldman** [comment]
> this looks good to me -- seems like 3.5 is having typing problems

Sorry about that -- will I never learn?  If this works, and @lucianopaz likes it, I propose to merge (with a little interactive rebasing to clean up).

**Comment by ColCarroll** [comment]
I'll wait for his input - we squash and merge by default, so no big worries about cleaning up the commits.

**Comment by rpgoldman** [comment]
> I'll wait for his input - we squash and merge by default, so no big worries about cleaning up the commits.

Squash and merge is fine, but if you want to see the bug I was trying to fix, we might want to squash into one commit introducing the new test, and one introducing the fix.  But if that's too fussy, one squashed commit is fine.

I just think it's important not to put garbage in the history by including my first try at a solution, which didn't work.

File: pymc3/model.py
Comment (lucianopaz) [inline pymc3/model.py]: Yes, you can add the methods dynamically inside `__new__` by appending them to the `dct`. Check out [this SO question](https://stackoverflow.com/q/32648767/7253166) and [this blog post on how to add them as bound method](https://www.ianlewis.org/en/dynamically-adding-method-classes-or-class-instanc).

File: pymc3/model.py
Comment (lucianopaz) [inline pymc3/model.py:268]: Nice!

File: pymc3/model.py
Comment (lucianopaz) [inline pymc3/model.py]:
```suggestion
        model = Model.get_context(error_if_none=False) # type: Optional['Model']
    if model is None:
        raise ValueError("No model on context stack.")
    return model
```
I would dedent and leave a single `return`

**Review by lucianopaz — APPROVED** [review]
Very nice work @rpgoldman! I left a few minor comments. Feel free to ignore them though. The only one that would require some work is to set `__enter__` and `__exit__` as bound methods in the metaclass' `__new__`.

File: pymc3/model.py
Comment (rpgoldman) [reply pymc3/model.py:268]: Thanks!

File: pymc3/model.py
Comment (rpgoldman) [reply pymc3/model.py]: I am making this change and will push it.

File: pymc3/model.py
Comment (rpgoldman) [reply pymc3/model.py]: A couple of questions:

1. I can't decide whether it's better to do this in `__new__` or `__init__`, or how I would make that decision.  Is there some reason it would be better to do this in `__new` than in `__init__`, the way I do?
2. Doing this in `__new__` by modifying the dict seems no nicer, syntactically, than what I do.  What I was looking for was a way to do this that would make my intentions clear to a maintainer reading my code afterwards.  Neither of these alternatives seems very nice, because they read as so different from adding a method in a normal class declaration.  In fact, I think messing with the dictionary might be even less readable than the assignment on lines 201 and 202 below.

Comment (lucianopaz) [reply pymc3/model.py]: I'm not sure if it's better to use `__new__` or `__init__` in our case. I think that using `__new__` should be preferred because the newly instantiated class will have all of its methods available when `__init__` is called. This could be useful if someone ever wants to write a new metaclass that inherits from `ContextMeta`, and shouldn't have to worry about when `supe().__init__(...)` gets called. Of course, I think that this is highly unlike to happen in our case, but I think that given that doing things in `__new__` or `__init__` is almost the same, we can just go for `__new__`.

Comment (rpgoldman) [reply pymc3/model.py]: OK, I'll make the change, then if the tests pass, I'll merge.  I'm kind of over-extended at work right now, so it might take me a little time.

**Comment by rpgoldman** [comment]
@ColCarroll Now that we have had our release (haven't we?), will you please merge this?

**Comment by rpgoldman** [comment]
If this passes, I've done the fix @lucianopaz requested, so I think this can be merged.

**Review by lucianopaz — APPROVED** [review]
Great work @rpgoldman!

**Pull request merged**
